src/control/character_controller.rs

```diff
@@ -267,6 +272,17 @@ impl KinematicCharacterController {
                     translation_remaining =
                         self.handle_slopes(&toi, &translation_remaining, &mut result);
                 }
+                if self.slide && self.reapply_lost_speed {
+                    let diff =
+                        pre_slide_translation_remaining.norm() - translation_remaining.norm();
+                    if let Some((normalized_dir, _)) =
+                        // threshold has to be > subtract_hit hardcoded correction, if not, we get made up movement in the direction of the hit normal
+                        UnitVector::try_new_and_get(translation_remaining, 1.0e-4)
+                    {
+                        // reapply the lost speed (but in the corrected direction)
+                        translation_remaining += *normalized_dir * diff;
+                    }
+                }
```

Comment:
That condition should probably be in case there is no stairs ? otherwise we might end up intersecting with the stair. (and it might result in not going up the stair, or worse an infinite loop ?)

Reply:
indeed, but even with the reapply in the `if !self.handle_stairs`, reapplying the speed can be slightly flaky, especially in 3d

im turning the PR back into a draft and looking into it